ksql-engine/src/test/java/io/confluent/ksql/function/FunctionValidatorTest.java

```diff
@@ -0,0 +1,339 @@
+    expectedException.expectMessage(
+        "Return type does not match supplied schema: MAP<STRING, INTEGER>. "
+            + "Type mismatch on map key type. "
+            + "expected: class java.lang.String, but got: class java.lang.Integer");
+
+    // When:
+    validator.validateReturnType(actualType, expectedType);
+  }
+
+  @Test
+  public void shouldThrowOnMapValueTypeMismatch() {
+    // Given:
+    final Type actualType = getReturnType("mapWithIntValue");
+
+    final SqlType expectedType = SqlTypes.map(SqlTypes.BIGINT);
+
+    // Then:
+    expectedException.expect(KsqlException.class);
+    expectedException.expectMessage(
+        "Return type does not match supplied schema: MAP<STRING, BIGINT>. "
+            + "Type mismatch on map value type. "
+            + "expected: class java.lang.Long, but got: class java.lang.Integer");
+
+    // When:
+    validator.validateReturnType(actualType, expectedType);
+  }
+
+  @Test
+  public void shouldNotThrowIfMapTypeMatchesSchema() {
+    // Given:
+    final Type actualType = getReturnType("mapWithIntValue");
+
+    final SqlType expectedType = SqlTypes.map(SqlTypes.INTEGER);
+
+    // When:
+    validator.validateReturnType(actualType, expectedType);
+
+    // Then:
+    // passed validation
+  }
+
+  @Test
+  public void shouldNotThrowOnRawMap() {
+    // Given:
+    final Type actualType = getReturnType("rawMap");
+
+    final SqlType expectedType = SqlTypes.map(SqlTypes.INTEGER);
+
+    // When:
+    validator.validateReturnType(actualType, expectedType);
+
+    // Then:
+    // passed validation
+  }
+
+  @Test
+  public void shouldThrowOnWrongRawType() {
+    // Given:
+    final Type actualType = getReturnType("rawMap");
+
+    final SqlType expectedType = SqlTypes.array(SqlTypes.INTEGER);
+
+    // Then:
+    expectedException.expect(KsqlException.class);
+    expectedException.expectMessage(
+        "Return type does not match supplied schema: ARRAY<INTEGER>. "
+            + "expected: interface java.util.List, but got: interface java.util.Map");
+
+    // When:
+    validator.validateReturnType(actualType, expectedType);
+  }
+
+  @Test
+  public void shouldThrowOnWildcard() {
+    // Given:
+    final Type actualType = getReturnType("withWildcard");
+
+    final SqlType expectedType = SqlTypes.array(SqlTypes.INTEGER);
+
+    // Then:
+    expectedException.expect(KsqlException.class);
+    expectedException.expectMessage(
+        "Wildcards, i.e. '?', are not supported. Please use specific types.");
+
+    // When:
+    validator.validateReturnType(actualType, expectedType);
+  }
+
+  @Test
+  public void shouldThrowOnTypeVar() {
+    // Given:
+    final Type actualType = getReturnType("typeVariable");
+
+    final SqlType expectedType = SqlTypes.array(SqlTypes.INTEGER);
+
+    // Then:
+    expectedException.expect(KsqlException.class);
+    expectedException.expectMessage(
+        "Type variables, e.g. V, are not supported. Please use specific types.");
```

Comment:
similarly, this isn't true. see `ToArray` UDF.

Reply:
Is this calling out the same thing as the previous comment?  If so, good to know ;)

Reply:
Basically, with slight nuance - the previous one says that it's okay to have top-level generics (i.e. `<T> T foo()`_ while this one says it's okay to have parameterized types (`<T> List<T> foo()`)